gemini/sample-apps/accelerating_product_innovation/app/pages_utils/utils_resources_store_embeddings.py

```diff
@@ -0,0 +1,477 @@
+    # (content and metadata) extracted from the document
+
+    data_packet = {}
+    data_packet["file_name"] = file_name
+    data_packet["page_number"] = str(page_number)
+    data_packet["chunk_number"] = str(chunk_number)
+    data_packet["content"] = file_content
+    return data_packet
+
+
+async def add_type_col(pdf_data: pd.DataFrame) -> pd.DataFrame:
+    """
+    Adds 'type' column to pdf_data.
+    """
+    pdf_data["types"] = pdf_data["content"].apply(get_type)
+    return pdf_data
+
+
+async def process_embedding(text: str) -> np.array:
+    """
+    Fetches results from embedding model.
+
+    Args:
+        text: Text to be stored as embeddings.
+    """
+    result = embedding_model_with_backoff([text])
+    return result
+
+
+async def generate_embeddings(pdf_data: pd.DataFrame) -> np.array:
+    """Generates the embeddings.
+
+    This function generates the embeddings.
+    It uses the 'map' function to apply the 'process_embedding' function to
+    each row of the 'content' column and returns the resulting numpy array.
+
+    Args:
+        pdf_data (pd.DataFrame): The PDF data.
+
+    Returns:
+        np.array: The embeddings for the PDF data.
+    """
+    tasks = [asyncio.create_task(process_embedding(x)) for x in pdf_data["content"]]
+    embeddings = await asyncio.gather(*tasks)
+    return np.array(embeddings)
+
+
+async def add_embedding_col(pdf_data: pd.DataFrame) -> pd.DataFrame:
+    """Adds an 'embedding' column to the PDF data.
+
+    This function adds an 'embedding' column to the PDF data.
+    It uses the 'apply' function to apply the 'generate_embeddings' function
+    to each row of the 'content' column and returns the resulting DataFrame.
+
+    Args:
+        pdf_data (pd.DataFrame): The PDF data.
+        ti (int): The task index.
+
+    Returns:
+        pd.DataFrame: The PDF data with the 'embedding' column.
+    """
+    # Make request data payload
+    json_data = pdf_data["content"].to_json()
+    data = {"pdf_data": json_data}
+    data_json = json.dumps(data)
+    # Make request headers
+    headers = {"Content-Type": "application/json"}
+
+    async with aiohttp.ClientSession() as session:
+        # URL for cloud function.
+        url = f"""https://us-central1-{PROJECT_ID}.cloudfunctions.net/text-embedding"""
+
+        # Call cloud function to generate embeddings with data and headers.
+        async with session.post(
+            url, data=data_json, headers=headers, verify_ssl=False
+        ) as response:
+            logging.debug("Inside IF else of session")
+
+            # Process cloud function Response
+            if response.status == 200:
+                response_text = await response.text()
+                final_response = json.loads(response_text)
+                embedding = final_response["embedding_column"]
+                pdf_data["embedding"] = embedding
+            else:
+                print("Request failed:", await response.text())
```

Comment:
I've seen quite a bit of this code in other places. Is there an opportunity to bring out into a function to reduce duplication?